### ISSUE TITLE
Optional whitespace control for well-formed result (not a duplicate)

### ISSUE THREAD

**Issue by TiGR** [twigphp/Twig#1005]
I know, there were a lot of issues about whitespace control, but I think that I've found a solution that might satisfy everyone. So, the rules for special whitespace removal are:
1. It is optional, might be enabled easily.
2. It removes only whitespace (no newlines) **before** control structures.
3. Only if this is just a single control structure (or maybe comment also) on line.
4. And only if there is nothing after it.

For example:

```
{% if good %}
        <hr />
{% endif %}
```

Would produce exactly the same result as

```
    {% if good %}
        <hr />
    {% endif %}
```

That is,

```
        <hr />
```

Isn't it logical? That is, lines with single control structure are ignored, no matter if there are spaces in the beginning or there aren't.

The problem is, that currently, in some cases there are just no way in Twig to achieve this without breaking readability of templates. Let's say you have something like this (a bit simplified code from real project, "set menu" added for easy reproduction):

```
{% set menu = {'home': {'url': '/', 'title': 'Home'},'more': {'title': 'More', 'items': [{'url': '/', 'title': 'foo'}, {'url': '/', 'title': 'bar'}]}, 'about': {'url': '/', 'title': 'About'}} %}
<nav>
    <ul>
    {% for id, item in menu %}
      {% if item.items %}
        <li class="dropdown">
            <a href="#">{{- item.title }}</a>
            <ul>
            {% for subitem in item.items %}
                <li><a href="{{ subitem.url }}">{{ subitem.title }}</a></li>
            {% endfor %}
            </ul>
        </li>
      {% else %}
        <li><a href="{{ item.url }}">{{ item.title }}</a></li>
      {% endif %}
    {% endfor %}
    </ul>
</nav>
```

You'd like to produce something beautiful, like this:

```
<nav>
    <ul>
        <li><a href="/">Home</a></li>
        <li class="dropdown">
            <a href="#">More</a>
            <ul>
                <li><a href="/">foo</a></li>
                <li><a href="/">bar</a></li>
            </ul>
        </li>
        <li><a href="/">About</a></li>
    </ul>
</nav>
```

But there is just **no simple way** to do this with Twig. You're not able to achieve this with spaceless or with '-'. Or you end up in putting '-' in weirdest places without any evident logic, or mess up everything with spaceless. That's not how template engine should let you produce well-formed result.

However, you could:
1. Manually remove spaces before all control structures. And voila! you've got the result you wished, but have broken template readability.
2. Move control structures opening tags to the beginning of the lines, increasing spaces within them correspondingly. This might work, but it breaks auto-indent in all editors.
3. Comment out spaces before tags.

I've written simple [Twig Preprocessor loader](https://gist.github.com/TiGR/5002699), so that I could easily strip all whitespace before control structures and comments and it all worked out just perfectly. As result Twig just ignores lines having only one control structure - no matter how many spaces are there before it (or there are none).

So, my proposal is to add this into twig somehow as an option.

**Comment by klor** [comment]
Inspiration:
[Whitespace Control](http://blog.rodneyrehm.de/archives/16-Smarty-Whitespace-Control.html) - a plug-in for Smarty:

* {-tag}  - remove white space infront of tag up to the previous non-whitespace character or beginning of the line
* {--tag}  - remove white space infront of tag up to the previous non-whitespace character
* {+-tag}, {-+tag}  - replace white space infront of tag up to the previous non-whitespace character by a single line-break
* {tag-}  - remove white space after tag up to the next non-whitespace character or end of the line
* {tag--}  - remove white space after tag up to the next non-whitespace character
* {tag+-}, {tag-+}  - replace white space after tag up to the next non-whitespace character by a single line-break
* {tag+}  - replace white space after tag up to the end of the line with an additional line-break

**Comment by fabpot** [comment]
fixed by #2925
